pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,281 @@
+		close(ev.eventResults)
+		q.closeWaitGroup.Done()
+
+		return ev.eventResults
+	default:
+		// The events channel may be closed even if an event has been pushed
+		// onto the events channel, as events are consumed off of the events
+		// channel asynchronously! If the EventQueue is closed before this
+		// event is processed, then it will be cancelled.
+
+		ev.stats.waitEnqueue.Start()
+		q.events <- ev
+		ev.stats.waitEnqueue.End(true)
+		ev.stats.waitConsumeOffQueue.Start()
+		return ev.eventResults
+	}
+}
+
+func (ev *Event) printStats() {
+	log.WithFields(logrus.Fields{
+		"eventType":                    reflect.TypeOf(ev.Metadata).String(),
+		"eventHandlingDuration":        ev.stats.durationStat.Total(),
+		"eventEnqueueWaitTime":         ev.stats.waitEnqueue.Total(),
+		"eventConsumeOffQueueWaitTime": ev.stats.waitConsumeOffQueue.Total(),
+	}).Debug("EventQueue event processing statistics")
+}
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
+// handling queued events is contained within this function. If the Metadata
+// within an Event does not implement the EventHandler interface, the function
+// will log an error, and the Event will not be processed. If the event queue is
+// closed, then all events which were queued up, but not processed, are
+// cancelled. Any event which is currently being processed will not be
+// cancelled. It is assumed that the caller handles both cases for an Event
+// (cancel, or result of event) gracefully.
+func (q *EventQueue) Run() {
+	go q.eventQueueOnce.Do(func() {
+		for ev := range q.events {
+			// Nil event is zero-value of *Event. Indicates q.events was closed.
+			// We can exit the queue.
+			if ev == nil {
+				return
+			}
+
+			q.closeWaitGroup.Done()
+			select {
+			case <-q.drain:
+				ev.stats.waitConsumeOffQueue.End(false)
+				close(ev.cancelled)
+				close(ev.eventResults)
+				ev.printStats()
+			default:
+				ev.stats.waitConsumeOffQueue.End(true)
+				ev.stats.durationStat.Start()
+				ev.Metadata.Handle(ev.eventResults)
+				// Always indicate success for now.
+				ev.stats.durationStat.End(true)
+				// Ensures that no more results can be sent as the event has
+				// already been processed.
+				ev.printStats()
+				close(ev.eventResults)
+			}
+		}
+	})
+}
+
+// Stop stops any further events from being processed by the EventQueue. Any
+// event which is currently being processed by the EventQueue will continue to
+// run. All other events waiting to be processed, and all events that may be
+// enqueued will not be processed by the event queue; they will be cancelled.
+// If the queue has already been stopped, this is a no-op.
+func (q *EventQueue) Stop() {
+	select {
+	case <-q.close:
+		log.Warning("tried to close event queue, but it already has been closed")
+	default:
+		q.closeOnce.Do(func() {
```

Comment:
non blocker: this `closeOnce` needs to be performed on line 244, otherwise with multiple calls made on `Stop()` only a few would print `tried to close event queue, but it already has been closed`

Reply:
I did this intentionally, as I figured users should be aware they are trying to close the queue multiple times, but wanted to avoid the panic scenario you outlined before.

Reply:
Fixed in latest upload.